---
title: "Class 16: Pathway Analysis from RNA-Seq Results"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

>> Live code: See the live code version here https://tinyurl.com/class16-bggn213

## 1. Differential Expression Analysis

Let's setup our R Studio with the program and the files that we'll be analyzing today.

>> Download and read input data

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703
The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1.

```{r include=FALSE}

library(DESeq2)

```

```{r}

metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak

colData = read.csv(metaFile, row.names=1)
head(colData)
```


```{r}

# Import countdata
rawcountData = read.csv(countFile, row.names=1)
head(rawcountData)


```

```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(rawcountData[,-1])
head(countData)
```

```{r}
dim(countData)

# This tells you nrow + ncol as a single function.
```

There are genes that have zero counts in all samples (i.e. columns). We should remove these.

First, to identify these columns we can use `rowsum`. 

Exclude these zero sum rows.

```{r}

validcountData <- countData[ rowSums(countData) != 0 , ]
head(validcountData)

```

```{r}
dim(validcountData)
```

## 2. Running DESeq2

```{r}
library(BiocVersion)
```

```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
dds
```

```{r}
res = results(dds)
```

```{r}
summary(res)
```

Derp derp derp, the plot looks like a volcano plot because it's a scatterplot. Volcano isn't *actually* a type of plot. It's a scatter plot that looks like a volcano ayyyy.

You take the -log in order to be able to visualize your p-value.

```{r}
plot( res$log2FoldChange, -log(res$padj) )

# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- ((res$padj < 0.01) & (abs(res$log2FoldChange) > 2 ))
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col= mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )

```

## 3. Annotate your genes.

To add annotation to the gene list, install some bioconductor packages.
However, I DO NOT DO THIS IN MY document I install them in the CONSOLE!!!

Let's map our ENSEMBLE gene ids to the more conventional gene SYMBOL.


```{r}

# BiocManager::install("AnnotationDbi")
# BiocManager::install("org.Hs.eg.db")
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys= row.names(res), 
                    keytype="ENSEMBL",
                    column= "SYMBOL",
                    multiVals="first")

head(res)


res$entrez = mapIds(org.Hs.eg.db,
                    keys= row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

head(res)


res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype= "ENSEMBL",
                    column= "GENENAME",
                    multiVals="first")

head(res, 10)
```

While working, occasionally save your file as a CSV in case something terrible happens to your R Studio instance.

```{r}
# write.csv(res, file="deseq_results.csv")
```


## 4. Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
# Run in your R console (i.e. not your Rmarkdown doc!)
# BiocManager::install( c("pathview", "gage", "gageData") )

```

Load the packages

```{r include=FALSE}
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Our input will be a vector of fold change values with ENTREZ gene ids as names

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r include=FALSE}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)
```

```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

Use pathview to visualize the pathways!

```{r}
# (gene.data=foldchanges, pathway.id="hsa04110")
```

```{r}
# A different PDF based output of the same data
# pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

```{r}
knitr::include_graphics("hsa04110.pathview.png")
```

We can look at a few more pathways.

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

```{r}
# pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

```{r echo=TRUE}
knitr::include_graphics(c("hsa04740.pathview.png", "hsa04640.pathview.png", "hsa00140.pathview.png", "hsa04630.pathview.png", "hsa04976.pathview.png"))
```









T-t-t-t-t-that's all folks!

```{r echo=FALSE}
knitr::include_graphics("github_signoff.gif")
```

